### ISSUE TITLE
Force a translation reload for the current language after an upgrade of the application

### ISSUE THREAD

**Issue by claudius76** [robisim74/angular-l10n#200]
Hi @robisim74 ,

I'm submitting a question that probably can help a lot of people. I would like to force a translation reload for the current language after an upgrade of the application. There is a simple way to do that?
I'm concerned about the following scenarios:
- Some translations were changed in an existing JSON file
- Some new translation files were added from the last version
The caching parameter is set to true now.
I tried to analyze the source code but all the methods that can help me to do this are private and I cannot use them.

I'm using angular-l10n with Angular 5.

Thanks

```
[] bug report
[] feature request
[x] support request
```

**Expected behavior**

**Actual behavior**

**Steps to reproduce the behavior**

StackBlitz Template: https://stackblitz.com/edit/angular-l10n

**Specifications: Angular version, library version, environment, browser**

**Comment by robisim74** [comment]
Hi @claudius76,

I get it. It is possible to change translation files at runtime. The problem is pure pipes: they are updated only if the key or language changes (passed as a parameter).

Make proposals if you have solutions.

**Comment by claudius76** [comment]
Hi @robisim74 ,

Thanks for the response.
Can you please provide me a piece of Typescript code that reloads the translation?
If I can have this I can try to make a proposal.

Regards,
Claudiu

**Comment by robisim74** [comment]
Sure.

The method that loads the translations is `loadTranslation`: https://github.com/robisim74/angular-l10n/blob/master/src/services/translation.service.ts#L188

The public method that you can use to load translations is `init`: https://github.com/robisim74/angular-l10n/blob/master/src/services/translation.service.ts#L72

The condition to change to force the loading of translation without switching language is: https://github.com/robisim74/angular-l10n/blob/master/src/services/translation.service.ts#L196

**Comment by claudius76** [comment]
@robisim74 ,

loadTranslation method is private in the TranslationService class so I cannot use it directly from the code.
I tried to use init in order to reload the translation as following:

`
reloadTranslation(retryOnFail = true) {
    let translationConfig = this.translationService.getConfiguration();
    translationConfig.caching = false;

    this.translationService.init().then(() => {
        translationConfig.caching = true;
        window.location.reload(true);
      }
      ,
      (error) => {
        alert(`Error reloading translation: ${error}`);

        if (retryOnFail)
          this.reloadTranslation(false);
        else {
          translationConfig.caching = true;
          window.location.reload(true);
        }
      });
  }
`

Do you think this method will reload properly the translations?

Thanks,
Claudiu

**Comment by robisim74** [comment]
As I said above, just call `init`, but it is also necessary to change the library at this line (otherwise you must change language): https://github.com/robisim74/angular-l10n/blob/master/src/services/translation.service.ts#L196

But as I said, the pure pipes will not update anyway.

**Comment by claudius76** [comment]
Hi @robisim74 ,

Thanks for the suggestions. They helped me to solve the problem.
I created a class that handles the translation reload.
Please check the code and tell me if I missed something.

`import { Injectable } from '@angular/core';
import { LocaleService, TranslationService, TranslationConfig } from 'angular-l10n';
import { LocalizationService } from './localization.service';

@Injectable()
export class TranslationManagerService {
  constructor(private translation: TranslationService,
    private localization: LocalizationService) {
  }

  public reloadTranslation() {
    let translationConfig = this.translation.getConfiguration();
    // Get the current locale
    let locale = this.localization.locale;
    // Get all the available languages
    let availableLanguages = locale.getAvailableLanguages();
    // Get the current language
    let crtLanguage = locale.getCurrentLanguage();
    // Remove the current language from the array to get the other languages
    let otherLanguages = availableLanguages.filter(lang => lang !== crtLanguage);

    if (otherLanguages && otherLanguages.length > 0) {
      // Change the language to force translation reload on init
      locale.setCurrentLanguage(otherLanguages[0]);
      // Disable the translation cache to force translation reload on init
      translationConfig.caching = false;

      // Reload the translation for the new language
      this.translation.init().then(() => {
        // Set the current language to the original value and reload translation
        this.loadTranslation(locale, translationConfig, crtLanguage);
      },
      () => {
        // Set the current language to the original value and reload translation
        this.loadTranslation(locale, translationConfig, crtLanguage);
      });
    } else {
      // Reload the page
      window.location.reload(true);
    }
  }

  private loadTranslation(locale: LocaleService, translationConfig: TranslationConfig, language: string) {
    // Set the current language
    locale.setCurrentLanguage(language);

    // Reload the translation
    this.translation.init().then(() => {
      // Enable the translation cache
      translationConfig.caching = true;
      // Reload the page
      window.location.reload(true);
    }
    ,
    () => {
      // Enable the translation cache
      translationConfig.caching = true;
      // Reload the page
      window.location.reload(true);
    });
  }
}`

I have a suggestion to make. It's a little bit annoying to change the language even temporarily to force translation reload. I think it would be nice to add a boolean parameter to the init function named forceReload with default value false. If forceReload is true, we add the forceReload parameter to loadTranslation function and change a bit the condition for that check like following:
`if (forceReload || (language != null && language != this.translation.getValue()))`
What do you think?

Thanks a lot,
Claudiu

**Comment by robisim74** [comment]
Please, don't get the instance of `LocaleService `in this way: `let locale = this.localization.locale;`. Instead, inject it in your class:

```
@Injectable()
export class TranslationManagerService {
  constructor(... private locale: LocaleService) {
}
```
Instead of adding another parameter, I was thinking of removing the condition (`language != this.translation.getValue()`): it was added only to avoid unexpected loading before caching was introduced.

Do you need this change also for Angular v5, right?

**Comment by claudius76** [comment]
Hi @robisim74 ,

1. The LocalizationService is a custom class that contains inside a public LocaleService member (please check the code below):

`@Injectable()
export class LocalizationService {
  constructor(public locale: LocaleService) {
  }
}`

The initialization that you provided will give me a different value than using this locale that it will be injected automatically by Angular?

2. I'm fine with your proposal of changing the condition by removing the language != this.translation.getValue() part.

3. Yes, I would like to have this change on Angular v5 if is possible.

Thank you for all support.

Claudiu

**Comment by robisim74** [comment]
Hi @claudius76,

I released the version of the library for Angular v5 that allows the reloading of the translation for the current language: https://github.com/robisim74/angular-l10n/releases/tag/v4.2.0

Could you let me have some feedback before I bring the same change to Angular v6?

Thank you.

**Comment by claudius76** [comment]
Hi @robisim74 ,

Sorry for the late response but I was in holiday.
Thanks for the fix. I will test it in the next days and I will send you my feedback.

I have a question related to the init function. Does it waits until all the translation files are loaded successfully or not?

Thank you,
Claudiu

**Comment by claudius76** [comment]
Hi @robisim74 ,

I tested the fix today with version 4.2.0 and it works. You can add the fix also to Angular 6 branch.

Thanks,
Claudiu

**Comment by claudius76** [comment]
Hi @robisim74 ,

Can you please add the fix for this on Angular 6 branch too?
We would like to upgrade the application to use Angular 6 and it will be nice to have this fix.

Thanks,
Claudiu

**Comment by robisim74** [comment]
Hi @claudius76,

It has already been added and it is available in the latest version.

Greetings

**Comment by claudius76** [comment]
Hi @robisim74 ,

I checked only the changes that were added in 5.2.0 release. After I saw your response, I noticed that the changes were included in the 5.1.0 version.
Congratulations for creating this awesome library that is very useful and easy to use.
Thanks for the feedback and all the support.

Claudiu
